airflow/www/api/experimental/endpoints.py

```diff
@@ -41,6 +47,124 @@
+    :return: List of DAG runs of a DAG with requested state,
+    """
+    state = request.args.get('state')
+    state_not_equal = request.args.get('state_not_equal')
+    execution_date_before = request.args.get('execution_date_before')
+    execution_date_after = request.args.get('execution_date_after')
+    dag_id = request.args.get('dag_id')
+
+    dagruns = get_all_dag_runs(dag_id=dag_id, state=state, state_not_equal=state_not_equal,
+                               execution_date_before=execution_date_before,
+                               execution_date_after=execution_date_after)
+
+    return jsonify(dagruns)
+
+
+@api_experimental.route('/task_instances', methods=['GET'])
+@requires_authentication
+def task_instances_filter():
+    """
+    Return the list of all dag_runs
+    :query param state: a query string parameter '?state=queued|running|success...'
+    :query param state_not_equal: a query string parameter '?state_not_equal=queued|running|success...'
+    :query param execution_date_before: a query string parameter to find all runs before provided date,
+    should be in format "YYYY-mm-DDTHH:MM:SS", for example: "2016-11-16T11:34:15".'
+    :query param execution_date_after: a query string parameter to find all runs after provided date,
+    should be in format "YYYY-mm-DDTHH:MM:SS", for example: "2016-11-16T11:34:15".'
+    :query param dag_id: String identifier of a DAG
+    :query param task_id: String identifier of a task
+    :return: List of task instances
+    """
+    state = request.args.get('state')
+    state_not_equal = request.args.get('state_not_equal')
+    execution_date_before = request.args.get('execution_date_before')
+    execution_date_after = request.args.get('execution_date_after')
+    dag_id = request.args.get('dag_id')
+    task_id = request.args.get('task_id')
+
+    task_instances = get_all_task_instances(dag_id=dag_id, state=state, state_not_equal=state_not_equal,
+                                            execution_date_before=execution_date_before,
+                                            execution_date_after=execution_date_after, task_id=task_id)
+
+    return jsonify(task_instances)
+
+
+@api_experimental.route('/dags', methods=['GET'])
+@requires_authentication
+def get_all_dags():
+    """
+    Returns a list of Dags
+    :query param is_paused: a query string parameter '?is_paused=true|false'
+    :return: List of all DAGs
+    """
+    is_paused = request.args.get('is_paused')
+    dag_list = get_dags(is_paused)
+
+    return jsonify(dag_list)
+
+
+@api_experimental.route('/dags/<string:dag_id>', methods=['GET'])
+@requires_authentication
+def get_dag_info(dag_id):
+    """
+    Returns information for a single dag
+    """
+    try:
+        dag = get_dag(dag_id)
+    except AirflowException as err:
+        _log.info(err)
+        response = jsonify(error="{}".format(err))
+        response.status_code = 404
+        return response
+    return jsonify(dag)
```

Comment:
Good practice to initialize ```dag = None``` just before the try-except since the return statement depends on it, or to return the dag inside of the try-block while the final return statement can just return None.

Reply:
good point.